content/gotraining.slide

```diff
@@ -212,21 +224,265 @@ Execute with
+- Write great commit messages
+- Think about the reviewer and review your own PRs first
+
+
+* Recommendations for CD @ANZ (2/2)
+
+- *Lint* strictly `golangci-lint` `run`
+- Ensure *100%* *coverage*
+- Test (unit & system), lint & cover as part of *CI* (Continuous Integration)
+- *Version* all deployments (use [[https://semver.org/][semver]]; tag source, images etc.)
+- *Document* for newbies in the README.md (must!) and use godoc
+- ... Non-prod vs Prod ?
+- Enjoy your CD as last step after successful CI run on master
+
+
+* Protect `master` branch on GitHub
+
+	Settings -> Options:
+	[ ] Allow merge commits
+	[x] Allow squash merging
+	[ ] Allow rebase merging
+
+	Settings -> Branches -> [Add rule|Edit] `master`:
+	[x] Require pull request reviews before merging
+	[x] Dismiss stale pull request approvals when new commits are pushed
+	[x] Require status checks to pass before merging
+	[x] Require branches to be up to date before merging
+	[x] Include administrators
+
+
+* Commit messages
+
+.image img/xkcd_git_commit.png
+.caption [[https://xkcd.com/1296/]]
+
+
+* Commit messages - spot the difference
+
+[[https://chris.beams.io/posts/git-commit/]]
+
+	$ git log --oneline -5 --author cbeams --before "Fri Mar 26 2009"
+
+	e5f4b49 Re-adding ConfigurationPostProcessorTests after its brief removal in r814. @Ignore-ing the testCglibClassesAreLoadedJustInTimeForEnhancement() method as it turns out this was one of the culprits in the recent build breakage. The classloader hacking causes subtle downstream effects, breaking unrelated tests. The test method is still useful, but should only be run on a manual basis to ensure CGLIB is not prematurely classloaded, and should not be run as part of the automated build.
+	2db0f12 fixed two build-breaking issues: + reverted ClassMetadataReadingVisitor to revision 794 + eliminated ConfigurationPostProcessorTests until further investigation determines why it causes downstream tests to fail (such as the seemingly unrelated ClassPathXmlApplicationContextTests)
+	147709f Tweaks to package-info.java files
+	22b25e0 Consolidated Util and MutableAnnotationUtils classes into existing AsmUtils
+	7f96f57 polishing
+
+versus
+
+	$ git log --oneline -5 --author pwebb --before "Sat Aug 30 2014"
+
+	5ba3db6 Fix failing CompositePropertySourceTests
+	84564a0 Rework @PropertySource early parsing logic
+	e142fd1 Add tests for ImportSelector meta-data
+	887815f Update docbook dependency and generate epub
+	ac8326d Polish mockito usage
+
+
+* Recipe for great commit messages
+
+[[https://chris.beams.io/posts/git-commit/]]
+
+- Separate subject from body with a blank line
+- Limit the subject line to 60 characters
+- *Capitalize* the subject line
+- Do not end the subject line with a period
+- Use the *imperative* *mood* in the subject line
+- Wrap the body at 80 characters
+- Use the body to explain what and why vs. how
+- Use `git` `rebase` `-i` `COMMIT_HASH` to rework your commits if necessary
```

Comment:
rebase may need quite a bit of explanation if it is unfamiliar. both the `rebase -i` form to clean up a sequence of commits, and the other form of rebase to relocate branches atop new base revisions (or whatever the right terminology is)

Reply:
Agreed, thanks for pointing this out!

Reply:
Agreed.

Reply:
I've resisted pointing these things out as these are bullet point slides. They introduce something that will be presented/talked about which is where the explanation is done.

I guess there's a balance between what is needed to be explained vs too much information for slides. I don't know exactly where that balance is, but we should defer to Julia as the presenter to have what she is comfortable with presenting, and not panic because the slides are not (re)covering what we think they should.

What I'm saying to you, @juliaogris, is you should feel free to respond to any comments on these slides with something like "ack, i'll add that to my presenter notes" and resolve the conversation there - in the interests of expediency and sanity.

Reply:
LOL LOL LOL (adding it to the speaker notes) LOL LOL LOL LOL LOL LOL LOL